File: cpp/itmat1/itmat1.cc
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <fstream>
#include <string>
#include <sstream>
#include <gperftools/profiler.h>

// Function to multiply two matrices
std::vector<std::vector<double>> multiplyMatrices(const std::vector<std::vector<double>>& mat1, const std::vector<std::vector<double>>& mat2) {
    int rows1 = mat1.size();
    int cols1 = mat1[0].size();
    int cols2 = mat2[0].size();

    std::vector<std::vector<double>> result(rows1, std::vector<double>(cols2, 0));

    for (int i = 0; i < rows1; i++) {
        for (int j = 0; j < cols2; j++) {
            for (int k = 0; k < cols1; k++) {
                result[i][j] += mat1[i][k] * mat2[k][j];
            }
        }
    }

    return result;
}

int main() {
    std::fstream f("../data/matrix/matrix.in");

    if (!f.good()) {
        std::cerr << "double check file path\n";
        return 1;
    }

    std::string line;
    std::getline(f, line);
    std::istringstream dimensions(line);
    int m, n, l;
    dimensions >> m >> n >> l;

    std::vector<std::vector<double>> A(m, std::vector<double>(n, 0)), B(n, std::vector<double>(l, 0));
    for (int i = 0; i < m && std::getline(f, line); i++) {
        std::istringstream row(line);
        for (int j = 0; j < n; j++) {
            row >> A[i][j];
        }
    }
    for (int i = 0; i < n && std::getline(f, line); i++) {
        std::istringstream row(line);
        for (int j = 0; j < l; j++) {
            row >> B[i][j];
        }
    }

    ProfilerStart("prof/itmat1.cprof");

    std::vector<std::vector<double>> result = multiplyMatrices(A, B);

    ProfilerStop();

    // for (auto &row : result) {
    //     for (const auto &val : row) {
    //         std::cout << val << "\n";
    //     }
    // }

    f.close();
    
    return 0;
}

File: go/itmat1/itmat1.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"runtime/pprof"
	"strconv"
	"strings"
)

var (
	memprofile = flag.String("memprofile", "", "write memory profile to this file")
	cpuprofile = flag.String("cpuprofile", "", "write cpu profile to this file")
	infile     = flag.String("infile", "../data/matrix/matrix.in", "Matrices to multiply")
)

// Function to multiply two matrices
func multiplyMatrices(mat1, mat2 [][]float64) [][]float64 {
	rows1 := len(mat1)
	cols1 := len(mat1[0])
	cols2 := len(mat2[0])

	result := make([][]float64, rows1)
	for i := range result {
		result[i] = make([]float64, cols2)
	}

	for i := 0; i < rows1; i++ {
		for j := 0; j < cols2; j++ {
			for k := 0; k < cols1; k++ {
				result[i][j] += mat1[i][k] * mat2[k][j]
			}
		}
	}

	return result
}

func main() {
	flag.Parse()

	file, _ := os.Open(*infile)
	defer file.Close()

	scanner := bufio.NewScanner(file)

	scanner.Scan()
	dimensions := strings.Fields(scanner.Text())

	m, _ := strconv.Atoi(dimensions[0])
	n, _ := strconv.Atoi(dimensions[1])
	l, _ := strconv.Atoi(dimensions[2])

	A := make([][]float64, m)
	for i := range A {
		scanner.Scan()
		line := scanner.Text()
		row := strings.Fields(line)

		A[i] = make([]float64, n)
		for j := range A[i] {
			A[i][j], _ = strconv.ParseFloat(row[j], 64)
		}
	}

	B := make([][]float64, n)
	for i := range B {
		scanner.Scan()
		line := scanner.Text()
		row := strings.Fields(line)

		B[i] = make([]float64, l)
		for j := range B[i] {
			B[i][j], _ = strconv.ParseFloat(row[j], 64)
		}
	}

	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		if err != nil {
			log.Fatal(err)
		}
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}

	C := multiplyMatrices(A, B)

	for r := range C {
		for _, val := range C[r] {
			fmt.Printf("%d\n", int(val))
		}
	}
}
